### ISSUE TITLE
ML model (Prototype)

### ISSUE THREAD

**Issue by DsP-0127** [DsP-0127/PranGanga#1]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

data = {
    'State': ['ANDHRA PRADESH'] * 6 + ['DAMAN AND DIU'] * 10 + ['GOA'] * 20 + ['GUJARAT'] * 10,
    'Min_Temperature': [25, 25, 22, 22, 22, 23, 24.9, 26.9, 26.8, 26.8, 26.8, 27.3, 27.4, 26, 27.1, 27, 27, 27, 27, 27],
    'Max_Temperature': [28, 28, 29, 30, 28, 29, 31.6, 31.8, 31.4, 32.7, 32.1, 32.5, 32.8, 35, 32.1, 30, 31, 31, 30, 30],
    'Min_Dissolved_O2': [6.2, 5.5, 4.5, 5.5, 5.8, 5.7, 6.2, 5.1, 4.2, 6.1, 6.2, 5.1, 4.1, 0.6, 5.9, 4.9, 6, 6.7, 4.8, 5.4],
    'Max_Dissolved_O2': [7.7, 8.3, 8.5, 8, 8.5, 8.5, 6.8, 6.2, 4.6, 6.6, 7.1, 5.8, 4.9, 6.9, 6.6, 7.7, 8.1, 8, 7.4, 7.7],
    'Min_pH': [7.1, 7.2, 7.2, 7.2, 7.3, 7.4, 7.2, 7.1, 7.5, 7.2, 7.6, 7.7, 7.3, 7.4, 7.2, 6.5, 6, 6.3, 6.7, 6.9],
    'Max_pH': [8.7, 8.4, 8.5, 8.3, 8.8, 8.4, 7.4, 7.7, 8, 7.5, 8, 8, 7.9, 8.5, 8, 7.7, 8, 7.8, 7.4, 8.3],
    'Min_Conductivity': [215, 185, 244, 251, 163, 164, 159, 20100, 18374, 183, 172, 746, 4400, 280, 216, 85, 35, 44, 42, 28],
    'Max_Conductivity': [365, 416, 415, 445, 355, 1435, 192, 21960, 21268, 245, 232, 1216, 5456, 14097, 287, 32460, 75, 52340, 6162, 112],
}

lengths = {key: len(value) for key, value in data.items()}
print("Lengths of each data column:", lengths)

if len(set(lengths.values())) == 1:
    df = pd.DataFrame(data)
else:
    print("Error: All arrays must be of the same length.")
   
if len(set(lengths.values())) == 1:
       X = df[['Min_Temperature', 'Max_Temperature', 'Min_Dissolved_O2', 'Max_Dissolved_O2', 'Min_pH', 'Max_pH', 'Min_Conductivity', 'Max_Conductivity']]
    y = df['Max_Temperature']  # Example target; adjust as needed

    
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    
    model = LinearRegression()
    model.fit(X_train, y_train)

    
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)

    print(f'Mean Squared Error: {mse}')

    
    coef_df = pd.DataFrame({'Feature': X.columns, 'Coefficient': model.coef_})
    print(coef_df.sort_values(by='Coefficient', ascending=False))

    plt.figure(figsize=(12, 6))
    plt.subplot(3, 1, 1)
    plt.plot(df['Min_Temperature'], label='Min Temperature (°C)', marker='o')
    plt.plot(df['Max_Temperature'], label='Max Temperature (°C)', marker='o')
    plt.title('Temperature over Different States')
    plt.ylabel('Temperature (°C)')
    plt.xticks(rotation=45)
    plt.legend()
    plt.grid()

    plt.subplot(3, 1, 2)
    plt.plot(df['Min_Dissolved_O2'], label='Min Dissolved O2 (mg/L)', marker='o')
    plt.plot(df['Max_Dissolved_O2'], label='Max Dissolved O2 (mg/L)', marker='o')
    plt.title('Dissolved Oxygen Levels over Different States')
    plt.ylabel('Dissolved O2 (mg/L)')
    plt.xticks(rotation=45)
    plt.legend()
    plt.grid()

    plt.subplot(3, 1, 3)
    plt.plot(df['Min_pH'], label='Min pH', marker='o')
    plt.plot(df['Max_pH'], label='Max pH', marker='o')
    plt.title('pH Levels over Different States')
    plt.ylabel('pH Level')
    plt.xticks(rotation=45)
    plt.legend()
    plt.grid()

    plt.tight_layout()
    plt.show()
